### PR TITLE
feat(finish): Require a fast-forward with --ff-only

### LINKED ISSUE #210: Add --ff-only to finish to require a fast-forward merge
`finish` can allow a fast-forward (`--ff`, the default) or force a merge commit (`--no-ff`), but there is no way to *require* a fast-forward and fail when the parent has moved on. This spec adds `--ff-only` as a precondition on the topic/parent relationship, checked before any merging starts. Refs #203.

### Goal

Give a release process a way to guarantee that what lands on the parent is bit-for-bit what was tested: if the parent has moved since the topic branch was cut or last synced, `finish` stops and says so instead of silently producing a merge commit whose tree differs from the tested tip. The guarantee is narrow and deliberate — it detects the condition, it does not resolve it.

### Expected Behavior

`--ff-only` is a **precondition**, not a fourth merge strategy. It is evaluated against the topic/parent relationship before anything is checked out or merged, so its meaning is the same whichever strategy the branch type uses.

The gate passes when the parent is an ancestor of the topic branch — including the case where they are equal. It fails whenever the parent carries a commit the topic does not, because then the parent tip after finishing would not be the tested topic tip. This covers both true divergence (each side has unique commits) and the parent-ahead-only case (the topic has nothing the parent lacks). Note that `git merge --ff-only` would *succeed* in the parent-ahead-only case, reporting "already up to date" — the gate is deliberately stricter, because "already up to date" means the tested tip never landed.

`--ff`, `--no-ff` and `--ff-only` are three values of one setting, mutually exclusive:

| Value | Behavior |
| --- | --- |
| `--ff` (default) | Fast-forward when possible, otherwise create a merge commit |
| `--no-ff` | Always create a merge commit |
| `--ff-only` | Fast-forward, or fail without touching the repository |

Success — the parent has not moved, so the merge is a true fast-forward:

```
$ git flow release finish 1.0.0 --ff-only
Switched to branch 'main'
Merging using strategy: merge
Created tag '1.0.0'
Updating child base branch 'develop' from 'main'...
Successfully finished branch 'release/1.0.0' and updated 1 child base branches
```

Failure — a hotfix landed on `main` that is not in `release/1.0.0`:

```
$ git flow release finish 1.0.0 --ff-only
Error: cannot fast-forward 'main' to 'release/1.0.0'
'main' has commits that are not in 'release/1.0.0'. --ff-only requires that what
lands on 'main' is exactly the tested branch tip. Integrate 'main' into
'release/1.0.0' and re-test, then finish again.
```

Exit code is 6 (validation error).

**Where the gate runs, and what "nothing is mutated" means.** The gate runs after the configured fetch and sync preflight — so it judges the parent state that would actually be merged — and before every step that changes the repository, including the pre-finish hook. On failure: no branch is checked out, no ref is moved, no tag is created, the topic branch is not deleted, no merge-state file is written, the pre-finish hook does not run, and the working tree is untouched. A configured fetch will already have updated remote-tracking refs before the gate; that is the one permitted side effect and is not part of the guarantee.

Scope of the constraint — **the upstream merge only** (topic into parent). It does not apply to the child auto-update step: back-merging `main` into `develop` is essentially never fast-forwardable, since `develop` carries commits `main` does not, so applying it there would break every release finish by construction.

Strategy interactions:

- **`--squash` is rejected** — a squash always creates a new commit, so a fast-forward is impossible by definition. Combining the two is a usage error (exit code 2), whether the squash comes from a flag, from `gitflow.<type>.finish.squash`, or from the branch type's Layer 1 `upstreamStrategy`.
- **`--rebase` is accepted.** When the gate passes, the topic is already a descendant of the parent and the rebase is a no-op; when the gate fails, `finish` aborts before anything is rewritten. `--ff-only` never rebases on the user's behalf to make the merge possible — rebasing onto a moved parent produces a tree that differs from the tested tip too, which is the very thing the flag exists to prevent.

Configuration follows the standard Layer 2 + CLI precedence, per branch type:

```
git config gitflow.release.finish.ff-only true
```

A CLI flag always wins over configuration, in either direction: `--ff` or `--no-ff` overrides a configured `ff-only`, and `--ff-only` overrides a configured `no-ff`. Within Layer 2, where the three keys can all be set independently, `ff-only = true` wins over both `no-ff` and `ff`. When `ff-only` is absent, the existing `no-ff` / `ff` resolution is unchanged.

### Test Scenarios

Baseline for all scenarios is `git flow init --defaults`. Where a scenario needs commits on two branches, each side changes a **different file**, so an ordinary merge succeeds and only the fast-forward condition is under test.

**Topologies.** *FF-capable*: `release start 1.0.0`, commit `release.txt` on the release branch, leave `main` untouched. *Diverged*: `release start 1.0.0`, commit `release.txt` on the release branch, then check out `main` and commit `main.txt` there. *Parent-ahead-only*: `release start 1.0.0` with no commits on it, then check out `main` and commit `main.txt`. Unless a scenario says otherwise, check out `release/1.0.0` again before invoking finish, so that a stray checkout to the parent is detectable.

1. FF-capable, flag — capture the release tip beforehand; `release finish 1.0.0 --ff-only` — succeeds; `main` equals the captured release tip; no merge commit was added to `main`; `1.0.0^{commit}` resolves to that same commit. (The `develop` back-merge does create a merge commit — `Repo.Merge` always passes `--no-ff` — so the assertion is about `main` only.)
2. Diverged, flag — diverged topology, `release finish 1.0.0 --ff-only` — fails with exit code 6; the error names both `main` and `release/1.0.0` and states that `main` has commits the release does not.
3. Parent-ahead-only — parent-ahead topology, `release finish 1.0.0 --ff-only` — fails with exit code 6, because `main`'s tip would not be the tested release tip even though `git merge --ff-only` alone would report "already up to date".
4. Nothing mutated on failure — before the failing command in scenario 2, capture the current branch, every branch SHA, the tag list and `git status`; afterwards all are identical, HEAD is still `release/1.0.0`, `release/1.0.0` still exists, and no merge-state file exists.
5. Pre-finish hook does not run on failure — install a pre-finish hook that writes a marker file, run the failing command from scenario 2 — the marker file does not exist.
6. FF-capable, config — `git config gitflow.release.finish.ff-only true`, FF-capable topology, `release finish 1.0.0` with no flags — succeeds; `main` equals the captured release tip.
7. Diverged, config — same config, diverged topology, `release finish 1.0.0` — fails with exit code 6.
8. `--ff` overrides configured ff-only — `gitflow.release.finish.ff-only true`, diverged topology, `release finish 1.0.0 --ff` — succeeds and creates a merge commit on `main`.
9. `--no-ff` overrides configured ff-only — same config, FF-capable topology, `release finish 1.0.0 --no-ff` — succeeds and creates a merge commit on `main` even though a fast-forward was possible.
10. `--ff-only` overrides configured no-ff — `gitflow.release.finish.no-ff true`, FF-capable topology, `release finish 1.0.0 --ff-only` — succeeds; `main` equals the captured release tip with no merge commit.
11. Layer 2 `ff-only` wins over Layer 2 `no-ff` — set both `gitflow.release.finish.ff-only true` and `gitflow.release.finish.no-ff true`, diverged topology, `release finish 1.0.0` — fails with exit code 6 rather than creating a merge commit.
12. `--ff-only --squash` rejected — FF-capable topology, `release finish 1.0.0 --ff-only --squash` — fails with exit code 2; the error names both options; the postconditions of scenario 4 hold.
13. Configured squash plus `--ff-only` rejected — `gitflow.release.finish.squash true`, `release finish 1.0.0 --ff-only` — same rejection and postconditions, confirming the check sees the resolved strategy rather than only the flags.
14. Layer 1 squash strategy plus `--ff-only` rejected — `git config gitflow.branch.release.upstreamStrategy squash`, `release finish 1.0.0 --ff-only` — same rejection, confirming the check covers a strategy originating at Layer 1.
15. `--no-ff --ff-only` rejected — FF-capable topology, `release finish 1.0.0 --no-ff --ff-only` — fails with exit code 2; the postconditions of scenario 4 hold.
16. `--ff --ff-only` rejected — FF-capable topology, `release finish 1.0.0 --ff --ff-only` — fails with exit code 2; the postconditions of scenario 4 hold.
17. `--ff-only --rebase`, gate passes — FF-capable topology, `release finish 1.0.0 --ff-only --rebase` — succeeds; `main` equals the captured release tip with no merge commit.
18. `--ff-only --rebase`, gate fails — diverged topology; capture `git rev-list release/1.0.0` beforehand; `release finish 1.0.0 --ff-only --rebase` — fails with exit code 6 and the captured revision list is unchanged, so no rebase was persisted.
19. Child auto-update is not constrained — FF-capable topology plus a commit to `develop.txt` on `develop` so the back-merge cannot fast-forward, `release finish 1.0.0 --ff-only` — succeeds; the `main` into `develop` back-merge creates a merge commit as usual.
20. Feature branch, diverged — commit `develop.txt` on `develop` after cutting `feature/f1`, `feature finish f1 --ff-only` — fails with exit code 6, confirming the flag is not release-specific.
21. Hotfix, FF-capable — `hotfix start 1.0.1` with `main` untouched, commit, `hotfix finish 1.0.1 --ff-only` — succeeds; `main` equals the captured hotfix tip.
22. Per-type configuration — `gitflow.release.finish.ff-only true` only, with `develop` diverged from `feature/f1` — `feature finish f1` still succeeds, confirming the key applies only to its own branch type.
23. Topic with no new commits and an unmoved parent — `release start 1.0.0`, finish immediately with `--ff-only` — succeeds; parent and topic are equal, so the gate is trivially satisfied.
24. Shorthand form — while on `release/1.0.0` in the diverged topology, `git flow finish --ff-only` — fails with exit code 6, confirming the shorthand path constructs the option too.
25. `integrate` flags unchanged (regression guard) — on a fresh repo, `git flow integrate develop --no-ff` still creates a merge commit and `git flow integrate develop --ff` still fast-forwards when possible.
26. `integrate` does not gain ff-only (regression guard) — `git flow integrate develop --ff-only` is rejected as an unknown flag, and `git config gitflow.develop.integrate.ff-only true` has no effect on an integrate that would otherwise create a merge commit.

### Out of Scope

- **Rebasing on the user's behalf** to make a non-fast-forwardable finish succeed. The gate reports the condition; resolving it is the user's decision.
- **Applying ff-only to the child auto-update step, `integrate`, or `update`.** This spec covers the upstream merge performed by `finish`.
- **A Layer 1 branch-type property.** `ff` / `no-ff` / `ff-only` are one tri-state setting and resolve at Layer 2, which is already per branch type; splitting one setting across layers would create ambiguity for no gain.
- **The existing `--ff --no-ff` combination**, which today silently resolves to `no-ff`. Unchanged; only combinations involving `--ff-only` are newly rejected.
- **The broader Layer 2 config key naming question**, tracked in #205. `ff-only` is chosen here for consistency with its siblings `ff` and `no-ff`.

### Technical Notes

- The gate needs a **read-only ancestry query**, which the codebase does not currently have — there is no `merge-base` or `is-ancestor` helper anywhere in `internal/git`. `repo.MergeFFOnly` (`internal/git/repo.go:535`) is a *mutating* merge and cannot serve as the pre-check: running it and reacting to its failure would already have checked out the parent, breaking the no-mutation guarantee, and it also succeeds in the parent-ahead-only case that scenario 3 requires to fail.
- `resolveFinishNoFF` (`internal/config/resolver.go:434`) currently resolves a single boolean from `gitflow.<type>.finish.no-ff` / `.ff`. It becomes a tri-state resolution over those two keys plus `ff-only`, and `MergeStrategyOptions.NoFF *bool` (`internal/config/resolver.go:70`) needs a representation carrying three values rather than a third independent boolean.
- **The resolver is shared with `integrate`.** `resolveMergeStrategy` is called with `"finish"` (`internal/config/resolver.go:88`) and with `"integrate"` (`:576`), and `integrate` exposes its own `--ff` / `--no-ff` (`cmd/integrate.go:260-261`) backed by `gitflow.<branch>.integrate.ff` / `.no-ff`. Widening the setting to a tri-state therefore changes a code path `integrate` depends on. `integrate` must behave exactly as it does today — no `--ff-only` flag, no `ff-only` key, no third state reaching its resolution. Guarded by scenarios 25 and 26.
- Flags are registered for the per-type form in `cmd/topicbranch.go:479-480`. The shorthand `git flow finish` shares those via `addFinishFlags` (`cmd/shorthand.go:185`) but builds its own `MergeStrategyOptions` at `cmd/shorthand.go:167`, so the new value has to be carried there too. (`cmd/shorthand.go:230` is the *integrate* shorthand and must not gain it.)
- Exit codes use the existing set in `internal/errors`: the gate failing is `ExitCodeValidationError` (6), conflicting options are `ExitCodeInvalidInput` (2). No new exit code is introduced.
- The squash conflict must be detected against the *resolved* strategy, since squash can arrive from `--squash`, `gitflow.<type>.finish.squash`, or Layer 1 `upstreamStrategy`.
- Documentation: `docs/git-flow-finish.1.md` (flag list, merge-strategy section, examples), `docs/gitflow-config.5.md` (the `gitflow.<type>.finish.*` key list), and `docs/avh-compatibility-analysis.md` — avh has `--[no]ff` but no `--ff-only`, so this is a beyond-avh extension.
- #170 (`--dry-run` for finish) should surface the ff-only gate in its plan output once both land.

### PR DESCRIPTION
**Pull request by alexrinass**
Adds `--ff-only` to `finish`, which requires the merge into the parent branch to be a fast-forward so that what lands on the parent is exactly the tested topic tip.

`--ff-only` is a precondition, not a fourth merge strategy: it is evaluated against the topic/parent relationship after the configured fetch and sync preflight but before any mutation, so a failure leaves no branch checked out, no ref moved, no tag created, no merge-state file written, and the pre-finish hook unrun. The same condition is re-checked immediately before the upstream merge, because the pre-finish hook runs in between and can move the parent. The gate passes when the parent is an ancestor of the topic (equal branches included) and fails whenever the parent carries a commit the topic does not — deliberately stricter than `git merge --ff-only`, which reports "already up to date" when the parent is merely ahead. `--ff`, `--no-ff` and `--ff-only` became three values of one tri-state setting (`config.FastForwardMode`) resolved through the standard Layer 2 + CLI precedence, backed by a new read-only `Repo.IsAncestor` helper and a `NotFastForwardableError` that exits 6. Combining `--ff-only` with `--ff`, `--no-ff`, or a squash strategy exits 2. Changes touch `cmd/finish.go`, `cmd/topicbranch.go`, `cmd/shorthand.go`, `internal/config/resolver.go`, `internal/git/repo.go`, and `internal/errors/errors.go`, with all 26 spec scenarios plus 12 edge cases and 4 regression guards covered in `test/cmd/finish_ffonly_test.go`, alongside the integrate regression guards.

Resolves #210

### Remarks

- The constraint applies to the upstream merge only. The child auto-update step still creates merge commits as usual, since back-merging `main` into `develop` is essentially never fast-forwardable and gating it would break every release finish.
- The precondition is enforced twice: once before anything is mutated, and once immediately before the merge. Only the second placement can catch a parent advanced by the pre-finish hook or a concurrent process, which would otherwise turn the guaranteed fast-forward into a silent merge commit. The late failure reports the same error and exit code and clears the merge state it would otherwise leave behind, but it is not a no-op — the hook has already run. A hook that advances the *topic* (the version-bump case) keeps the parent an ancestor and still fast-forwards.
- The upstream merge itself is run with Git's `--ff-only`, on both the merge and the post-rebase call site. Checking the topology beforehand cannot constrain what the merge then does: a repository with `merge.ff = false` would otherwise get a merge commit and a successful exit. The existing merge helpers gained the flag rather than being replaced, so message handling, `--no-verify` and the conflict classification are untouched, and `--continue` deliberately does not pass it.
- Both sides of the comparison are qualified as `refs/heads/*`. Git's revision parser resolves `refs/tags/<name>` first, so a tag sharing a branch's name would otherwise make the gate judge an object other than the branch the merge moves.
- `--rebase` is accepted: when the gate passes the rebase is a no-op, and when it fails nothing is rewritten. `--ff-only` never rebases on the user's behalf to make the merge possible.
- The squash conflict is checked against the *resolved* strategy, so a squash arriving from `--squash`, `gitflow.<type>.finish.squash`, or a Layer 1 `upstreamStrategy` is rejected alike.
- `integrate` shares the resolver but is unchanged: the `ff-only` config key is read in the `finish` namespace only, integrate registers no `--ff-only` flag, and its flag mapping cannot produce the third value. Scenarios 25 and 26 guard this.
- The `finish` manpage's EXIT STATUS table was factually wrong for the codes this change touches (it listed 2 as "Git operation failed" and 6 as "GPG signing failed"). Corrected against `internal/errors` rather than appending accurate clauses to inaccurate entries.

**Review focus:**
- `cmd/finish.go` — the squash conflict check, the gate's placement relative to the fetch preflight and `finishBranch`, the re-check at the top of `handleMergeStep`, and the parent-existence pre-check that preserves today's `BranchNotFoundError`.
- `internal/config/resolver.go:467-491` — tri-state resolution order: Layer 2 reads `.no-ff`, then `.ff`, then finish-only `.ff-only`, preserving the pre-existing ff-beats-no-ff rule when `ff-only` is absent.
- `cmd/shorthand.go` — all three fast-forward flags are read by value rather than through `getBoolPtr`, whose "any mention is a selection" rule made `--ff-only --no-ff=false` a usage error on this surface and an accepted command on the other. The shared `ffModeFromFlags` keeps the `--ff`/`--no-ff` precedence identical for both.

### CODE REVIEW

File: cmd/finish.go [hunk 1]
```diff
@@ -244,10 +253,43 @@ func executeFinish(repo *git.Repo, branchType string, name string, continueOp bo
...
+//
+// The parent's existence is verified here because the gate runs ahead of
+// finishBranch's own check, so a misconfigured parent keeps the error and exit
+// code it produces today instead of surfacing as a raw merge-base failure.
+func requireFastForwardable(repo *git.Repo, topic string, parent string) error {
+	if err := repo.BranchExists(parent); err != nil {
+		return &errors.BranchNotFoundError{BranchName: parent}
+	}
+
+	fastForwardable, err := repo.IsAncestor(parent, topic)
```
[the remark this answers is not in the document]
Comment (alexrinass) [reply cmd/finish.go]: Correct, and the mismatch was worse than "ambiguous": `requireFastForwardable` verifies the parent with `BranchExists`, which already qualifies as `refs/heads/<branch>`, and then compared whatever `merge-base --is-ancestor` resolved — potentially a different object than both the branch it just verified and the branch the merge goes on to move.

Fixed in 86839af by qualifying both sides at the call site: `refs/heads/`+parent and `refs/heads/`+topic. `IsAncestor` itself is left alone deliberately — it is a general-purpose ancestry helper that takes any revision (with unit tests covering unknown refs), so the branch semantics belong to the caller that knows it is comparing branches. The error message keeps the plain names the user typed.

Two regression tests in `test/cmd/finish_ffonly_test.go` (R1, R2): one with a tag named `develop` at a non-ancestor commit, which made the gate reject a finish that fast-forwards cleanly, and one with a tag named `feature/f1` at the parent's tip, which made it accept a finish it must reject. Both fail without the fix.

File: cmd/finish.go [hunk 2]
```diff
@@ -244,10 +253,43 @@ func executeFinish(repo *git.Repo, branchType string, name string, continueOp bo
...
 
+	// Enforce the --ff-only precondition. It runs here, after the fetch/sync
+	// preflight, so it judges the parent state that would actually be merged, and
+	// before finishBranch, which runs the pre-finish hook and writes the merge
+	// state file. --continue and --abort return well above this point, so a
+	// resumed operation never re-evaluates the gate.
+	if resolvedOptions.RequireFastForward {
+		if err := requireFastForwardable(repo, name, branchConfig.Parent); err != nil {
+			return err
+		}
```
[the remark this answers is not in the document]
Comment (alexrinass) [reply cmd/finish.go:265]: Agreed on the hole, with one distinction that shaped the fix: the two directions are not equally bad.

- A **topic** advanced between check and merge (a version-bump pre-finish hook, the common case) leaves the parent an ancestor. The merge is still a fast-forward and the parent still ends at the topic tip, so the guarantee holds and this must keep working.
- A **parent** advanced in between makes the merge a merge commit. That is exactly what `--ff-only` exists to prevent, and it is the real hole.

Fixed in d895943 with the second of your two suggestions: the ancestry is re-checked at the top of `handleMergeStep`, immediately before the merge, failing with the same `NotFastForwardableError` (exit 6). I did not take the "capture the commit and run `git merge --ff-only` against it, skipping the rebase path" route: it would change `--no-verify` handling and alter the rebase strategy's behavior, whereas re-validation covers merge and rebase uniformly and leaves the merge mechanics untouched.

Two details worth flagging:

- `--continue` is unaffected. `handleContinue` completes the merge itself and advances the step before re-entering the state machine, so `handleMergeStep` is only ever reached on the initial finish — a resumed operation still never re-evaluates the gate. `integrate` shares this step handler but can never resolve to the ff-only mode, so the check is inert there.
- A late failure is not the "nothing was mutated" abort the precondition gate offers: the pre-finish hook has already run and whatever it did stands. Nothing has been merged, though, so there is no operation to resume, and the merge state written moments earlier is cleared rather than left behind. The manpage now says this explicitly instead of implying the earlier guarantee covers both checks.

Tests (R3, R4): a pre-finish hook that advances the parent must fail with exit 6, leaving no merge commit on `develop` and no state file; a pre-finish hook that advances the topic must still succeed and fast-forward. The second one is the guard against the fix over-firing — it passed before the change and still passes after.

**Comment by alexrinass** [comment]
Round 2 came through as two suppressed comments on the review rather than inline threads, so both are answered here. Both were reproduced against the built binary before anything was changed, and both were correct.

> **cmd/finish.go:693** — This re-check still does not enforce the promised fast-forward. After it passes, the code executes a plain `git merge`, so a repository with `merge.ff=false` will create a merge commit even for this topology; a parent update between this check and that merge can do the same. `--ff-only` can therefore return success while violating its contract. The upstream merge itself must use `--ff-only` (against the qualified topic ref), including the post-rebase merge, and translate a non-fast-forward rejection to `NotFastForwardableError`; the early check can remain to preserve the no-mutation failure path.

Right, and the `merge.ff` half is the part no amount of pre-checking can fix. Reproduced on a repository with `merge.ff = false`: `git flow release finish 1.0.0 --ff-only` exited 0 and left `main` at a merge commit two parents deep instead of the release tip — the exact outcome the option promises to prevent.

Fixed in 44f9baf: the requirement is carried into the merge itself, on both the merge strategy's call site and the post-rebase one, where a command-line `--ff-only` overrides the `merge.ff` config. The existing merge helpers gained the flag instead of being replaced by `MergeFFOnly`, so the merge message handling, `--no-verify` and the conflict classification stay exactly as they were, and the rebase still runs — only its trailing merge is now constrained.

Two things I did not do, with reasons:

- **The pre-merge check stays.** It is not redundant with Git's enforcement: it produces `NotFastForwardableError` with the remedy rather than "Not possible to fast-forward, aborting", and it fails before the parent branch is even checked out. Both layers are cheap.
- **No error mapping from Git's rejection back to `NotFastForwardableError`.** Reaching it now requires a ref to move inside the window between the re-check and the merge, which cannot be provoked deterministically in a test. I would rather not add a branch that no test can reach; the outcome there is a clear failure with a non-zero exit, not a silent violation.

`--continue` deliberately does not pass the flag: a resumed operation never re-evaluates the gate, and an ff-only finish cannot reach that path anyway, since the gate makes its rebase a no-op and leaves no conflict to resume from.

Tests: R5 (`merge.ff=false` with `--ff-only` fast-forwards) and R6 (the same with `--rebase`, covering the post-rebase call site). Both fail without the fix.

> **cmd/shorthand.go:172** — `getBoolPtr` only checks whether a flag was mentioned and returns a hard-coded value, so explicit false values are misread here. For example, shorthand `finish --ff-only --no-ff=false` is rejected as a conflict, while the per-type command reads the actual values and accepts it. Read all three Cobra boolean values before calling `ffModeFromFlags` so both finish surfaces agree.

Confirmed exactly as described: `git flow feature finish f1 --ff-only --no-ff=false` exits 6 with the gate error, while `git flow finish --ff-only --no-ff=false` exits 2 with "cannot combine --ff-only with --no-ff". The same rule also made `--ff=false` select fast-forward instead of leaving the decision to config.

Fixed in 79db5e9 by reading all three flags by value. The `--ff`/`--no-ff` precedence already lives in the shared `ffModeFromFlags`, so every ordinary spelling behaves exactly as before — only the explicit `=false` forms change, and they change to match the per-type surface. R7 covers it.

Full suite green locally after both changes.

File: cmd/finish.go [hunk 3]
```diff
@@ -650,19 +735,19 @@ func handleMergeStep(repo *git.Repo, cfg *config.Config, state *mergestate.Merge
...
 	case strategySquash:
 		mergeErr = repo.MergeSquashWithMessage(state.FullBranchName, resolvedOptions.SquashMessage, resolvedOptions.NoVerify)
 	case strategyMerge:
 		if resolvedOptions.MergeMessage != "" {
 			expandedMsg := util.ExpandMessagePlaceholders(resolvedOptions.MergeMessage, state.FullBranchName, state.ParentBranch)
-			mergeErr = repo.MergeWithMessage(state.FullBranchName, expandedMsg, resolvedOptions.NoFastForward, resolvedOptions.NoVerify)
+			mergeErr = repo.MergeWithMessage(state.FullBranchName, expandedMsg, resolvedOptions.NoFastForward, resolvedOptions.RequireFastForward, resolvedOptions.NoVerify)
 		} else {
-			mergeErr = repo.MergeWithOptions(state.FullBranchName, resolvedOptions.NoFastForward, resolvedOptions.NoVerify)
+			mergeErr = repo.MergeWithOptions(state.FullBranchName, resolvedOptions.NoFastForward, resolvedOptions.RequireFastForward, resolvedOptions.NoVerify)
```
[the remark this answers is not in the document]
Comment (alexrinass) [reply cmd/finish.go]: Accepted and fixed in 78af2cf. The gate and the merge could resolve different objects, which is exactly the inconsistency the earlier fix was meant to remove.

I qualified the topic ref only when `--ff-only` is in effect, not at every finish. Git echoes the ref it was given into the default merge message, so `git merge refs/heads/feature/x` commits as `Merge branch 'refs/heads/feature/x'` (verified locally). Under `--ff-only` the merge writes no commit at all, so the qualification is free there; applying it to every finish would change the subject of every merge commit git-flow creates. The squash call site is unreachable under `--ff-only` — the combination is rejected up front — so it keeps the bare name.

That leaves the same shadowing hazard for non-ff-only finishes, where merging the tag produces a wrong-but-committed merge rather than a silent no-op. Noted for a follow-up issue: fixing it properly means passing the qualified ref together with an explicit message that reproduces git's current wording.

Covered by `TestFinishFFOnlyMergesTopicBranchNotTag`, which fails on the previous code — develop stayed where it was while the feature branch was tagged away and deleted.

**Comment by alexrinass** [comment]
### Round 3 suppressed comments

Both accepted, fixed in 5ff1075 (git layer) and 78af2cf (finish). The inline comment on the merge call sites is answered in its own thread.

**Rebase can rewrite the topic onto a moved parent** (`cmd/finish.go:740`). Correct, and it breaks the rule the whole feature rests on: `--ff-only` never rebases on the user's behalf. The rebase is now skipped whenever `--ff-only` is in effect, rather than the window being narrowed — when the topology is valid the rebase is a no-op, and when it is not, the finish has to be reported rather than rebased into shape. `TestFinishFFOnlyWithRebaseDoesNotRebaseOntoMovedParent` drives it deterministically with a post-checkout hook that advances the parent between the pre-merge re-check and the merge: on the previous code the finish succeeded with a rewritten topic, now it exits 6 with `feature/f1`'s revision list unchanged. `TestFinishFFOnlyWithRebaseKeepsTopicHistoryOnSuccess` holds scenario 17 in place — `--ff-only --rebase --keep` still succeeds and the topic keeps its exact commits.

**Git's own ff-only rejection is misclassified** (`internal/git/repo.go:601`). Also correct, and the consequence is what decided it: exit 3 instead of 6, merge state left on disk, and the parent still checked out. The merge helpers now wrap a `git.ErrNotFastForward` sentinel, and `handleMergeStep` maps it to `NotFastForwardableError`, clears the merge state and returns to the topic branch — the same end state as the pre-merge abort.

On detection: instead of matching `Not possible to fast-forward, aborting.`, the classifier re-derives the condition from the refs — when `--ff-only` was requested and the merged ref is not a descendant of `HEAD`, the merge cannot have been a fast-forward, whatever else went wrong. That survives a translated git, which the sentence does not. The earlier round's objection that this path could not be tested no longer holds either: the post-checkout hook moves the parent inside the window, so `TestFinishFFOnlyReportsGitRefusalAsGateFailure` exercises it end to end (exit 6, no merge state, back on the topic branch), and `TestMergeWithOptionsFFOnlyRefusalWrapsSentinel` / `TestMergeWithMessageFFOnlyRefusalWrapsSentinel` pin the classification at the git layer.

Manpage and changelog updated for the rebase suppression and the late refusal. Full suite green locally.

File: internal/git/repo.go [hunk 4]
```diff
@@ -565,60 +593,79 @@ func (r *Repo) MergeFFOnly(branch string) error {
...
+		return r.classifyMergeFailure(branchName, ffOnly, string(output))
+	}
 
-		if len(conflictOutput) > 0 ||
-			strings.Contains(outputStr, "Automatic merge failed") ||
-			strings.Contains(outputStr, "CONFLICT") ||
-			strings.Contains(outputStr, "merge failed") ||
-			strings.Contains(outputStr, "needs merge") {
-			return fmt.Errorf("merge conflict: %s", outputStr)
+	return nil
```
[the remark this answers is not in the document]
Comment (alexrinass) [reply internal/git/repo.go:657]: Good catch, and it matters more than its rarity suggests: parent-ahead-only is precisely the asymmetry the gate was built for (spec scenario 3), and it was the one topology the late merge path could not see.

Fixed in 6b439d4. Both helpers now call a shared `verifyFastForwardLanded` after a `--ff-only` merge git reported as successful; it resolves `HEAD` and the merged ref to commits (`^{commit}`, so an annotated tag cannot compare unequal to the commit it names) and returns the existing `git.ErrNotFastForward` sentinel on a mismatch. It sits next to `classifyMergeFailure`, so the success and failure paths of a `--ff-only` merge are classified in one place. `handleMergeStep`'s existing `errors.Is(mergeErr, git.ErrNotFastForward)` branch then applies unchanged — exit 6, merge state cleared, topic branch checked back out — which the regression test asserts rather than assumes.

Regression `TestFinishFFOnlyRejectsParentMovedAheadOfTopic` uses the post-checkout hook technique: the hook fast-forwards `main` onto `release/1.0.0` and then commits on top, inside the window between the pre-merge re-check and the merge. It asserts the hook really produced a parent strictly ahead of the topic (so it cannot pass vacuously), then exit 6, the gate error rather than a generic merge failure, `release/1.0.0` still present at its original tip, no tag created, HEAD back on the topic, and no merge state file. On the pre-fix code it fails with the finish succeeding: `Created tag '1.0.0'` on the hook's commit and the topic deleted. Three unit tests in `test/internal/git/repo_test.go` cover the helper directly, including the equal-branches control.

The over-firing risk is covered: equality is the success case, so `--ff-only` on branches that are already the same commit (spec scenario 23) still succeeds — verified explicitly, along with scenarios 17 and 18. Full suite green.
